src/work/BasicWork.h

```diff
@@ -0,0 +1,153 @@
+// Copyright 2018 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+#pragma once
+
+#include "main/Application.h"
+#include <set>
+
+namespace stellar
+{
+
+class Application;
+
+/** BasicWork is an implementation of a finite state machine,
+ * that is used for async or long-running tasks that:
+ *  - May need to be broken into steps, so as not to block the main thread
+ *  - May fail and need to retry, after some delay
+ *
+ *  BasicWork manages all state transitions via `crankWork` and `setState`.
+ *  It also supports a retry mechanism internal to BasicWork.
+ *  While customers can trigger cranking and check on BasicWork's status,
+ *  it is implementers' responsibility to implement `onRun`, that
+ *  hints the next desired state.
+ */
+
+class BasicWork : public std::enable_shared_from_this<BasicWork>,
+                  private NonMovableOrCopyable
+{
+  public:
+    static size_t const RETRY_NEVER;
+    static size_t const RETRY_ONCE;
+    static size_t const RETRY_A_FEW;
+    static size_t const RETRY_A_LOT;
+    static size_t const RETRY_FOREVER;
+
+    // Publicly exposed state of work:
+    // WORK_RUNNING - work has been created and is currently in progress
+    // (note that this might mean that work is being destructed also).
+    // Implementers return RUNNING when work needs to be scheduled to run more.
+    // WORK_WAITING - work should not be scheduled to run, as it's waiting
+    // for some event (process exit, retry timeout, etc)
+    // Implementers return WAITING when they don't want the work to be
+    // scheduled. Note that it is implementers' responsibility then to correctly
+    // use `wakeUp` to exit WAITING state. WORK_SUCCESS - terminal successful
+    // state WORK_FAILURE - terminal unsuccessful state. When this state is
+    // returned, depending on the Work's retry strategy, either a retry will be
+    // scheduled or work will cease execution.
+    enum class State
+    {
+        WORK_RUNNING,
+        WORK_WAITING,
+        WORK_SUCCESS,
+        WORK_FAILURE,
+    };
+
+    BasicWork(Application& app, std::string name, size_t maxRetries);
+    virtual ~BasicWork();
+
+    std::string const& getName() const;
+    virtual std::string getStatus() const;
+    State getState() const;
+    bool isDone() const;
+
+    size_t getMaxRetries() const;
+    uint64_t getRetryETA() const;
+
+    // Main method for state transition, mostly dictated by `onRun`
+    void crankWork();
```

Comment:
I think it would be nice to also remove this from public interface. It is used only in classes derivered from BasicWork itself (Work and WorkScheduler). As it is used through shared_ptrs it may not be as trivial as just making it protected, but it may be worth it.